### ISSUE TITLE
False-positive need-check-nil calling assert(loadfile(...)) after replacing _G.loadfile

### ISSUE THREAD

**Issue by lewis6991** [EmmyLuaLs/emmylua-analyzer-rust#1259]
## Description

After replacing the global `loadfile`, calling the result of `assert(loadfile(...))` produces a false-positive `need-check-nil`. The same expression also produces a contradictory `unnecessary-assert` diagnostic.

Encountered in Neovim, where `vim.loader` can replace `_G.loadfile`. The following standalone reproduction needs no Neovim dependencies or files to load.

## Minimal reproduction

Create an otherwise empty directory containing `.emmyrc.json` with `{}` and `repro.lua`:

```lua
---@param path string
---@return function?, string?
local function load_chunk(path)
  if path == '' then
    return nil, 'empty path'
  end
  return function() return 42 end
end
_G.loadfile = load_chunk

print(assert(loadfile('example.lua'))())
```

Run from that directory:

```sh
emmylua_check . --warnings-as-errors
```

Tested with `emmylua_check 0.25.1` on macOS arm64. Also reproduces with 0.23.2.

## Actual behavior

Exit status 1; both diagnostics point at line 11:

```text
warning: Unnecessary assert: this expression is always truthy [unnecessary-assert]
warning: function assert(loadfile('example.lua')) may be nil [need-check-nil]
```

## Expected behavior

`assert` either raises an error or returns a non-nil function, so calling its return value is safe. Running the reproduction with LuaJIT exits successfully and prints `42`.

Removing `_G.loadfile = load_chunk` and calling `assert(load_chunk('example.lua'))()` instead produces no diagnostics. The replacement of the built-in global appears relevant.

## Workaround

Replacing the last line with the following produces no diagnostics:

```lua
local loader, err = loadfile('example.lua')
assert(loader, err)
print(loader())
```

**Comment by lewis6991** [comment]
Further reduction: the replacement signature matters; replacing the global alone is not sufficient.

The builtin declares three optional parameters (`filename?: string`, `mode?: std.loadmode`, `env?: table`) and correlated return overloads (`function` or `nil, string`). The original reproduction used one required parameter and independent nullable returns.

A reproduction using Neovim's actual parameter/return annotations also produces both warning diagnostics:

```lua
---@param filename? string
---@param mode? "b"|"t"|"bt"
---@param env? table
---@return function?, string?
local function load_chunk(filename, mode, env)
  if filename == '' then
    return nil, 'empty path'
  end
  return function() return 42 end
end
_G.loadfile = load_chunk

print(assert(loadfile('example.lua'))())
```

Changing **only** `---@param mode? "b"|"t"|"bt"` to `---@param mode? std.loadmode` eliminates both warnings in 0.25.1 (only unused-parameter hints remain). The builtin `std.loadmode` alias itself is exactly `"b"|"t"|"bt"`.

Thus semantically equivalent alias vs inline-union parameter types affect this result. This appears related to how the replacement and builtin signatures are combined, rather than a general failure of `assert` to narrow nullable function returns. That is an observation, not a confirmed internal cause.
